targets/unity-v2/source/Shared/Public/Pipeline/OneDSEventPipeline.cs

```diff
@@ -0,0 +1,357 @@
+                    return false;
+                }
+
+                return true;
+            }
+            catch (Exception e)
+            {
+                logger.Error($"Exception in synchronous WriteEvent from {e.Source} with message: {e.Message}");
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Writes an event into the pipeline and returns a task that allows user to wait for a result
+        /// when this particular event is processed by pipeline and sent out to backend.
+        /// </summary>
+        /// <param name="request">The emit event request</param>
+        /// <returns>A task that allows user to wait for a result.</returns>
+        public async Task<IPlayFabEmitEventResponse> IntakeEventAsync(IPlayFabEmitEventRequest request)
+        {
+            try
+            {
+                this.ThrowIfDisposed();
+                var resultPromise = new TaskCompletionSource<IPlayFabEmitEventResponse>();
+
+                if (request == null)
+                {
+                    // We don't want to throw and break pipeline because of a bad event
+                    logger.Error("Request passed to event pipeline is null");
+                    resultPromise.SetCanceled();
+                }
+                else if (!this.isActive)
+                {
+                    logger.Warning("Event pipeline is not active");
+                    resultPromise.SetCanceled();
+                }
+
+                // Non-blocking add, return immediately and report a dropped event
+                // if event buffer is full or marked as complete for adding.
+                else
+                {
+                    ((PlayFabEmitEventRequest)request).ResultPromise = resultPromise;
+                    if (!this.eventBuffer.TryAdd(request))
+                    {
+                        logger.Error("Event buffer is full or complete and event {0} cannot be added", ((PlayFabEmitEventRequest)request).Event.Name);
+                        resultPromise.SetCanceled();
+                    }
+                }
+
+                return await resultPromise.Task;
+            }
+            catch (Exception e)
+            {
+                logger.Error($"Exception in IntakeEventAsync from {e.Source} with message: {e.Message}");
+                var taskCompletionSource = new TaskCompletionSource<IPlayFabEmitEventResponse>();
+                taskCompletionSource.SetResult(new PlayFabEmitEventResponse());
+                return await taskCompletionSource.Task;
+            }
+        }
+
+        private void Cancel()
+        {
+            this.eventBuffer?.CompleteAdding();
+
+            // This should also direct all stages to mark their output buffers as complete for adding:
+            this.pipelineCancellationTokenSource?.Cancel();
+        }
+
+        private void ThrowIfDisposed()
```

Comment:
I don't like using an exception when an if will do, and I VERY much do not like explicitly turning an if info an exception.
It serves no purpose other than making code harder to maintain, and introducing a bunch of unintended consequences.
Every place where this function is used, convert it to be an if-return.